Test/RadixSort/main.cpp

```diff
@@ -38,131 +39,166 @@ class Stopwatch
 			OrochiUtils::waitForCompletion();
+
 			sw.start();
-			m_sort->sort( srcGpu, dstGpu, testSize, 0, testBits, m_tempBuffer );
+			m_sort.sort( srcGpu, dstGpu, testSize, 0, testBits, m_tempBuffer );
 			OrochiUtils::waitForCompletion();
 			sw.stop();
 			float ms = sw.getMs();
-			float gKeys_s = testSize/1000.f/1000.f/ms;
-			printf("%3.2fms (%3.2fGKeys/s) sorting %3.1fMkeys\n", ms, gKeys_s, testSize/1000.f/1000.f);
+			float gKeys_s = testSize / 1000.f / 1000.f / ms;
+			printf( "%3.2fms (%3.2fGKeys/s) sorting %3.1fMkeys\n", ms, gKeys_s, testSize / 1000.f / 1000.f );
 		}
 
-		vector<u32> dst( testSize );
-		OrochiUtils::copyDtoH( dst.data(), dstGpu, testSize );
+		std::vector<u32> dstKey( testSize );
+		OrochiUtils::copyDtoH( dstKey.data(), dstGpu.key, testSize );
+
+		std::vector<u32> dstValue( testSize );
+		OrochiUtils::copyDtoH( dstValue.data(), dstGpu.value, testSize );
+
+		std::vector<u32> indexHelper( testSize );
+		std::iota( std::begin( indexHelper ), std::end( indexHelper ), 0U );
+
+		std::stable_sort( std::begin( indexHelper ), std::end( indexHelper ), [&]( const auto indexA, const auto indexB ) noexcept { return srcKey[indexA] < srcKey[indexB]; } );
+
+		const auto rearrange = []( auto& targetBuffer, const auto& indexBuffer ) noexcept
+		{
+			std::vector<u32> tmpBuffer( std::size( targetBuffer ) );
+
+			for( int i = 0; i < std::size( targetBuffer ); ++i )
+			{
+				tmpBuffer[i] = targetBuffer[indexBuffer[i]];
+			}
+
+			targetBuffer = std::move( tmpBuffer );
+		};
+
+		rearrange( srcKey, indexHelper );
+		rearrange( srcValue, indexHelper );
 
-		std::sort( src.begin(), src.end() );
 		for( int i = 0; i < testSize; i++ )
 		{
-			if( dst[i] != src[i] ) 
+			if( dstKey[i] != srcKey[i] || dstValue[i] != srcValue[i] )
 			{
-				printf( "fail\n" );
+				for( int j = ( i >= 5 ) ? i - 5 : 0; j < i + 5; ++j )
+				{
+					printf( "fail: [%d]: %d %d %d %d\n", j, dstKey[j], srcKey[j], dstValue[j], srcValue[j] );
+				}
+
 				__debugbreak();
 				break;
 			}
 		}
 
-		OrochiUtils::free( srcGpu );
-		OrochiUtils::free( dstGpu );
-		printf("passed: %3.2fK keys\n", testSize/1000.f);
+		OrochiUtils::free( srcGpu.value );
+		OrochiUtils::free( dstGpu.value );
+
+		OrochiUtils::free( srcGpu.key );
+		OrochiUtils::free( dstGpu.key );
+
+		printf( "passed: %3.2fK keys\n", testSize / 1000.f );
 	}
 
 	template<typename T>
 	inline T getRandom( const T minV, const T maxV )
 	{
 		double r = std::min( (double)RAND_MAX - 1, (double)rand() ) / RAND_MAX;
 		T range = maxV - minV;
-		return ( T )( minV + r * range );
+		return (T)( minV + r * range );
 	}
 
   private:
 	oroDevice m_device;
 	oroCtx m_ctx;
-	Oro::RadixSort* m_sort;
+	Oro::RadixSort m_sort;
 	u32* m_tempBuffer;
 };
 
 enum TestType
 {
-	TEST_SINGLE,// test single kernel sort
+	TEST_SINGLE, // test single kernel sort
 	TEST_SIMPLE,
 	TEST_PERF,
 	TEST_BITS,
 };
 
-int main(int argc, char** argv )
+int main( int argc, char** argv )
```

Comment:
Note: some of the formats are corrected automatically by clang-format when editing the file in Visual Studio.